x/feegrant/types/filtered_fee.go

```diff
@@ -0,0 +1,112 @@
+package types
+
+import (
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	proto "github.com/gogo/protobuf/proto"
+)
+
+var _ FeeAllowanceI = (*AllowedMsgFeeAllowance)(nil)
+var _ types.UnpackInterfacesMessage = (*AllowedMsgFeeAllowance)(nil)
+
+// UnpackInterfaces implements UnpackInterfacesMessage.UnpackInterfaces
+func (a *AllowedMsgFeeAllowance) UnpackInterfaces(unpacker types.AnyUnpacker) error {
+	var allowance FeeAllowanceI
+	return unpacker.UnpackAny(a.Allowance, &allowance)
+}
+
+// NewAllowedMsgFeeAllowance creates new filtered fee allowance.
+func NewAllowedMsgFeeAllowance(allowance FeeAllowanceI, allowedMsgs []string) (*AllowedMsgFeeAllowance, error) {
+	msg, ok := allowance.(proto.Message)
+	if !ok {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrPackAny, "cannot proto marshal %T", msg)
+	}
+	any, err := types.NewAnyWithValue(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AllowedMsgFeeAllowance{
+		Allowance:       any,
+		AllowedMessages: allowedMsgs,
+	}, nil
+}
+
+// GetAllowance returns allowed fee allowance.
+func (a *AllowedMsgFeeAllowance) GetAllowance() (FeeAllowanceI, error) {
+	allowance, ok := a.Allowance.GetCachedValue().(FeeAllowanceI)
+	if !ok {
+		return nil, sdkerrors.Wrap(ErrNoAllowance, "failed to get allowance")
+	}
+
+	return allowance, nil
+}
+
+// Accept method checks for the filtered messages has valid expiry
+func (a *AllowedMsgFeeAllowance) Accept(fee sdk.Coins, blockTime time.Time, blockHeight int64, msgs []sdk.Msg) (bool, error) {
+	if !a.allMsgTypesAllowed(msgs) {
+		return false, sdkerrors.Wrap(ErrMessageNotAllowed, "message does not exist in allowed messages")
+	}
+
+	allowance, err := a.GetAllowance()
+	if err != nil {
+		return false, err
+	}
+
+	return allowance.Accept(fee, blockTime, blockHeight, msgs)
+}
+
+func (a *AllowedMsgFeeAllowance) allMsgTypesAllowed(msgs []sdk.Msg) bool {
+	for _, msg := range msgs {
+		found := false
+		for _, allowedMsg := range a.AllowedMessages {
+			if allowedMsg == msg.Type() {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return false
+		}
+	}
```

Comment:
We actually want to charge extra gas for doing this.

Reply:
Also, this is two nested for loops - `O(m * n)` complexity - we don't want that. Instead we should build a map with the allowed messages in one loop, then in a second loop we check each message against the map - that's `O(m + n)` complexity which is much better IMHO.